text/0000-deprecate-cargo-categories.md

```diff
@@ -0,0 +1,131 @@
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Cargo supports two types of metadata when tagging the purpose/scope of a crate:
+
+1. Categories, which are restricted to a curated set of allowed categories.
+2. Keywords, which are unrestricted and user-defined.
+
+However, the reality is that the actual curation of categories is extremely loose, and the only process for adding a category is to open a PR for the crates.io repository. The required review is effectively asking for an "LGTM" reply from one of the crates.io team members, as seen in [the history for the `categories.toml` file on the repository](https://github.com/rust-lang/crates.io/commits/main/src/boot/categories.toml).
+
+Ultimately, the most important distinction between categories and keywords has really been that the crates.io team has control over the presentation of categories, although a more important distinction is that categories cannot be removed, only added.
+
+Recently, [a discussion](https://github.com/rust-lang/crates.io/discussions/6762) was opened in the crates.io repository on whether the cryptocurrencies category should be removed from crates.io, due to the plethora of issues surrounding them. This should not be treated as a reason for adopting the RFC (although it was a motivation to write it), but instead as something that brought up the fact that categories cannot be removed by policy.
+
+Because categories cannot be removed, they also cannot be renamed or otherwise curated by the community. However, by switching to keywords, we can effectively solve this problem; community members can simply start publishing their crates under different keywords and older, unsupported crates wouldn't have any issue remaining published under their older versions.
+
+To make the change backwards-compatible, any crates using categories in their manifests will implicitly add those strings to keywords instead. Ultimately, switching to a unified system solves a couple problems:
+
+1. Instead of being gatekept by a PR to the crates.io repository, categories can organically be adopted by community members in the form of keywords. Which keywords are most popular and useful can be decided organically.
+2. Keywords can still be given descriptions and other metadata on crates.io, although no distinction between these "special" keywords and other keywords is made in cargo itself. This allows making changes to the way crates are presented without having to worry about backwards compatibility.
+3. Adding, removing, and modifying the curated set of keywords is no longer a technical choice, but a cultural one.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Categories in `Cargo.toml` are now deprecated. Setting `categories` will now trigger a warning and suggest to use `keywords` instead. To ensure that crates still build correctly in these cases, the provided `categories` are implicitly converted into `keywords`.
```

Comment:
> To ensure that crates still build correctly in these cases, the provided `categories` are implicitly converted into `keywords`.

This makes it sound like by adding a deprecation warning, crates using categories will no longer be able to build, so cargo must move categories to keywords.

However, a warning should not affect building.  I can see crates.io doing the conversion on their side, since they have to deal with a variety of cargo versions.  I can also see us removing support for `package.categories` in an edition with `cargo fix` automatically migration people.  Those are both a little different from how I read this.

Reply:
Right, so, I definitely over-simplified in the guide-level explanation, and will try and take a look at it later to see how I can reword it.

I don't really think we should ever remove the categories keyword, since the edition system would force us to parse it anyway, and it's not like we really gain anything by forcing crates past a certain edition to not use it. We could make the theoretical lint that triggers deny-by-default, but we also don't need an edition bump to do that either.

I do think that the potential to `cargo fix` is on the table, though, assuming we go through with this.

Ultimately, we don't want to break old builds, so, the best we can do is implicitly convert to keywords and tell people to not use categories. Breaking builds would be breaking backwards compatibility, after all, and that's not allowed.

Reply:
Its fine for the guide to simplify things; its just the guide made it sound like this was the only place explaining a part of what is happening.

> I don't really think we should ever remove the categories from the manifest, since the edition system would force us to parse it anyway, and it's not like we really gain anything by forcing crates past a certain edition to not use it. We could make the theoretical lint that triggers deny-by-default, but we also don't need an edition bump to do that either.

If we are completely changing the interpretation of `categories` and deprecating it (warning + maybe cargo fix), then I think it is worth considering migrating people to avoid confusion ("I have the categories field, where does it show up?", "I'm looking at categories, but what does it mean?").

> Ultimately, we don't want to break old builds, so, the best we can do is implicitly convert to keywords and tell people to not use categories. Breaking builds would be breaking backwards compatibility, after all, and that's not allowed.

Except what is build-breaking that you are avoiding?  That was mentioned in the RFC and in the reply but there isn't any explanation as to what you are referring to.  Are you talking saying we can't remove the field altogether?  Or something else?  What made this more confusing is that the concern for "build breaking" is tied to the registries implicitly converting.